Fix win64 build on MSVC with CUDA+OpenCL

zao opened the pull request:
This pull request addresses all the problems I can see on Windows 10 + Visual Studio 2015 Win64 with the CUDA 8.0 SDK.

It renames all uses of HPX `get_gid` into `get_id` which seems to be a replacement, and it honors the internal protocol of `init_no_addref` in HPX `promise_base`.

It also gets rid of some annoying warnings and spurious use of `<unistd.h>` in tests.

This ought to resolve #38.

hkaiser commented:
LGTM, thanks!

diehlpk commented:
After merging into master and running it on circle-ci results into 

In file included from /usr/local/include/hpx/runtime/serialization/detail/preprocess.hpp:18:
/usr/local/include/hpx/lcos/local/promise.hpp:37:35: error: no matching constructor for initialization of 'shared_state_type' (aka 'hpx::opencl::lcos::detail::event_data<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::serialization::serialize_buffer<char, std::allocator<char> > >')
              : shared_state_(new shared_state_type(init_no_addref()), false)
                                  ^                 ~~~~~~~~~~~~~~~~
/usr/local/include/hpx/lcos/detail/promise_base.hpp:131:17: note: in instantiation of member function 'hpx::lcos::local::detail::promise_base<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::opencl::lcos::detail::event_data<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::serialization::serialize_buffer<char, std::allocator<char> > > >::promise_base' requested here
              : base_type()
                ^
/hpx/opencl/lcos/event.hpp:253:13: note: in instantiation of member function 'hpx::lcos::detail::promise_base<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::opencl::lcos::detail::event_data<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::serialization::serialize_buffer<char, std::allocator<char> > > >::promise_base' requested here
          : base_type()
            ^
/hpx/opencl/buffer.cpp:121:24: note: in instantiation of member function 'hpx::opencl::lcos::event<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::serialization::serialize_buffer<char, std::allocator<char> > >::event' requested here
    event<buffer_type> ev( device_gid );
                       ^
/hpx/opencl/lcos/event.hpp:83:11: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'init_no_addref' (aka 'hpx::lcos::detail::future_data_refcnt_base::init_no_addref') to 'const hpx::opencl::lcos::detail::event_data<hpx::serialization::serialize_buffer<char, std::allocator<char> >, hpx::serialization::serialize_buffer<char, std::allocator<char> > >' for 1st argument
    class event_data
          ^
/hpx/opencl/lcos/event.hpp:91:9: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
        event_data() {}
        ^
1 warning and 2 errors generated.

zao committed:
Fix build issues on MSVC and HPX master

The master branch of HPX appears to have deprecated `get_gid()` and has
hidden it behind a macro that is not defined for some configurations.
All uses of `get_gid()` for CUDA and core have been replaced with
`get_id()`.

FindNVRTC looks in insufficient numbers of directories for the Win64
library, this augments the search paths.

The examples pull in `<unistd.h>` without reason, remove it.

zao committed:
Add missing constructors for init_no_addref

HPX's `promise_base` and `promise_data` has a protocol where they pass
`init_no_addref` as an optimization to avoid overheads in reference
counting. To fix this HPXCL's `event_data` for the `event` type gains a
constructor to forward this call properly to the base class.
